src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -594,34 +587,28 @@ protected virtual void OnCustomCommand(int command)
         ///       contains multiple associated services. Loads the specified services into memory so they can be
         ///       started.</para>
         /// </devdoc>
-        public static void Run(ServiceBase[] services)
+        public static unsafe void Run(ServiceBase[] services)
         {
             if (services == null || services.Length == 0)
                 throw new ArgumentException(SR.NoServices);
 
-            int sizeOfSERVICE_TABLE_ENTRY = Marshal.SizeOf<SERVICE_TABLE_ENTRY>();
-
-            IntPtr entriesPointer = Marshal.AllocHGlobal(checked((services.Length + 1) * sizeOfSERVICE_TABLE_ENTRY));
+            IntPtr entriesPointer = Marshal.AllocHGlobal(checked((services.Length + 1) * sizeof(SERVICE_TABLE_ENTRY)));
+            Span<SERVICE_TABLE_ENTRY> entries = new Span<SERVICE_TABLE_ENTRY>((void*)entriesPointer, services.Length + 1);
+            entries.Clear();
             try
             {
-                SERVICE_TABLE_ENTRY[] entries = new SERVICE_TABLE_ENTRY[services.Length];
                 bool multipleServices = services.Length > 1;
-                IntPtr structPtr;
 
                 for (int index = 0; index < services.Length; ++index)
                 {
-                    services[index].Initialize(multipleServices);
-                    entries[index] = services[index].GetEntry();
-                    structPtr = entriesPointer + sizeOfSERVICE_TABLE_ENTRY * index;
-                    Marshal.StructureToPtr(entries[index], structPtr, fDeleteOld: false);
+                    ServiceBase service = services[index];
+                    service.Initialize(multipleServices);
+                    // This method allocates on unmanaged heap; Make sure that the contents are freed after use.
+                    entries[index] = service.GetEntry();
                 }
 
-                SERVICE_TABLE_ENTRY lastEntry = new SERVICE_TABLE_ENTRY();
-
-                lastEntry.callback = null;
-                lastEntry.name = (IntPtr)0;
-                structPtr = entriesPointer + sizeOfSERVICE_TABLE_ENTRY * services.Length;
-                Marshal.StructureToPtr(lastEntry, structPtr, fDeleteOld: false);
+                // The members of the last entry in the table must have NULL values to designate the end of the table.
+                entries[entries.Length - 1] = default;
```

Comment:
This assignment is unnecessary (the comment would be enough), but it is ok to keep it for clarity.